libcxx/include/__algorithm/ranges_contains_subrange.h

```diff
@@ -0,0 +1,145 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCPP___ALGORITHM_RANGES_CONTAINS_SUBRANGE_H
+#define _LIBCPP___ALGORITHM_RANGES_CONTAINS_SUBRANGE_H
+
+#include <__algorithm/ranges_starts_with.h>
+#include <__config>
+#include <__functional/identity.h>
+#include <__functional/ranges_operations.h>
+#include <__functional/reference_wrapper.h>
+#include <__iterator/concepts.h>
+#include <__iterator/distance.h>
+#include <__iterator/indirectly_comparable.h>
+#include <__iterator/projected.h>
+#include <__ranges/access.h>
+#include <__ranges/concepts.h>
+#include <__utility/move.h>
+
+#if !defined(_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER)
+#  pragma GCC system_header
+#endif
+
+#if _LIBCPP_STD_VER >= 23
+
+_LIBCPP_BEGIN_NAMESPACE_STD
+
+namespace ranges {
+namespace __contains_subrange {
+struct __fn {
+  template <input_iterator _Iter1,
+            sentinel_for<_Iter1> _Sent1,
+            input_iterator _Iter2,
+            sentinel_for<_Iter2> _Sent2,
+            class _Pred,
+            class _Proj1,
+            class _Proj2,
+            class _Offset>
+  static _LIBCPP_HIDE_FROM_ABI constexpr bool __contains_subrange_fn_impl(
+      _Iter1 __first1,
+      _Sent1 __last1,
+      _Iter2 __first2,
+      _Sent2 __last2,
+      _Pred& __pred,
+      _Proj1& __proj1,
+      _Proj2& __proj2,
+      _Offset __offset) {
+    if (__offset < 0)
+      return false;
+    else {
+      for (; __offset >= 0; __offset--, __first1++) {
+        auto result = ranges::starts_with(
+            std::move(__first1),
+            std::move(__last1),
+            std::move(__first2),
+            std::move(__last2),
+            std::ref(__pred),
+            std::ref(__proj1),
+            std::ref(__proj2));
+        if (result)
+          return true;
+      }
+      return false;
+    }
```

Comment:
is there a reason why you did not define in terms of `ranges::search` as in the standard?

Reply:
Yes.
Initially, I didn't aware that I can use `ranges::search`.
Then I tried to replace `ranges::starts_with` with `ranges::search` and I got 

> constraints not satisfied for class template 'subrange' [with _Iter = cpp20_input_iterator<int *>, _Sent = cpp20_input_iterator<int *>, _Kind = std::ranges::subrange_kind::unsized]

error in the test file.

I tried to test with forward iterator. It works well.

Reply:
`contains_subrange` only worked for `forward_iterator`, which is the same as `ranges::search`. Why do you test it with an `input_iterator`? It will fail as expected.